components/logins/src/migrate_sqlcipher_db.rs

```diff
@@ -67,40 +99,95 @@ fn sqlcipher_3_compat(conn: &Connection) -> Result<()> {
-    conn.execute("DETACH DATABASE plaintext", NO_PARAMS)?;
-    Ok(())
+//Manually copy over row by row from sqlcipher db to a plaintext db
+pub fn migrate_from_sqlcipher_db(
+    cipher_conn: &mut Connection,
+    new_db_store: LoginStore,
+    encryption_key: &str,
+) -> Result<MigrationMetrics> {
+    let start_time = Instant::now();
+
+    let mut metrics: MigrationMetrics = MigrationMetrics::default();
+
+    // Select From both LoginsL and LoginsM with a union to ensure we're covering our
+    // migration cases (one table has but not the other, etc)
+    let mut select_stmt = cipher_conn.prepare(&GET_BY_GUID_SQL)?;
+
+    // Use raw rows to avoid extra copying since we're looping over an entire table
+    let mut rows = select_stmt.query(NO_PARAMS)?;
+    while let Some(row) = rows.next()? {
+        metrics.num_processed += 1;
+        let guid: String = row.get("guid")?;
+        let username: String = row.get("username").unwrap_or_default();
+        let password: String = row.get("password")?;
+        // migrating hostname to the new column origin
+        let origin: String = row.get("hostname")?;
+        let http_realm: Option<String> = row.get("httpRealm")?;
+        // migrating formSubmitURL to the new column action origin
+        let form_action_origin: Option<String> = row.get("formSubmitURL")?;
+        let username_field: Option<String> = row.get("usernameField")?;
+        let password_field: Option<String> = row.get("passwordField")?;
+        let time_created: i64 = row.get("timeCreated")?;
+        let time_last_used: i64 = row.get("timeLastUsed").unwrap_or_default();
+        let time_password_changed: i64 = row.get("timePasswordChanged")?;
+        let times_used: i64 = row.get("timesUsed")?;
+
+        // TODO: Discuss
+        // Need to handle in loginsL: local_modified, is_deleted
+        // Feels like we potentially shouldn't migrate these fields?
+
+        // Need to handle in loginsM: is_overridden, server_modified
+        // Similar to the other: I only see server_modified potentially being needed
+
+        // encrypt the username/password data
+        let encryptor = EncryptorDecryptor::new(encryption_key)?;
+        let login: Login = Login {
+            id: guid.to_string(),
+            username_enc: encryptor.encrypt(&username)?,
+            password_enc: encryptor.encrypt(&password)?,
+            origin,
+            http_realm,
+            form_action_origin,
+            username_field: username_field.unwrap_or_default(),
+            password_field: password_field.unwrap_or_default(),
+            // TO_DO: Do we need to convert from microsecond timestamps to milliseconds??
+            time_created,
+            time_last_used,
+            time_password_changed,
+            times_used,
+        };
+
+        // Leveraging the add_or_update to get free fixup
+        match new_db_store.add_or_update(login) {
```

Comment:
I like the idea of leveraging `add_or_update()`, but I'm not sure that this will work if my comment above is correct and you need to insert data to both the loginsL and loginsM tables.

I'm also always paranoid of using shared code for migrations.  This means that if we ever want to change `add_or_update()` or the `Login` struct, then it's going to change how the migration works and it could be pretty easy to introduce a bug since we're no longer thinking about migrating these old databases.

Reply:
I definitely go back and forth on this:

On one hand - writing code once ensures that any new fixups/ways we add to the database stays consistent for anyone migrating from an old version but does potentially impede new functionality if we always have to mold to "legacy" DBs. However in having them completely separate means we still need to make sure any new changes need to be included in the APIs as well as the migration logic else we accidentally put in bad data....definitely need to think about this a little more!

Reply:
> I like the idea of leveraging `add_or_update()`, but I'm not sure

We might be able to leverage the fact that we know it's an empty DB - so it might mean that `add_or_update()` gets a little refactored and we still call the underlying implementation - the main thing I'm keen on is ensuring our real fixup/validation stuff is called and not using sql here for any inserts that might be necessary.

(I'm still leaning towards the nuclear option of changing/migrating the mirror TBH - I'm usually actively looking for incremental approaches to refactoring, but this is a bit of a rare case - but I don't think we need to make that call yet)